### ISSUE TITLE
Add support for pre restore MSBuild target that Visual Studio triggers

### ISSUE THREAD

**Issue by atanasa** [NuGet/Home#4781]
I'd like to have an MSBuild target that runs before a restore operation. Also this target should be triggered by both CLI and Visual Studio. Probably a post restore target would be good too.

I would use this to extend the restore logical operation to consume other package sources that are not supported by NuGet.

According to this documentation https://docs.microsoft.com/en-us/nuget/schema/msbuild-targets#target-build-order there should be one.
As discussed in dotnet/cli#5683 and #4634 there is a pre restore target that is private (`_GenerateRestoreProjectSpec`) but is not triggered by Visual Studio.

*Edit*:

Use `CollectPackageReferences` as per [NuGet.Targets](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets)

**Comment by emgarten** [comment]
@davkean any ideas on how we could run targets from Visual Studio that are set to run before restore?

**Comment by eerhardt** [comment]
MSBuild has extensibility built into it.  So you can add an MSBuild Target that runs before Restore by putting the following in your MSBuild .*proj file (or a .targets file imported by the .*proj):

```
  <Target Name="DoSomethingBeforeRestore"
          BeforeTargets="Restore">
    <Message Text="Got here before Restore!" Importance="high" />
  </Target>
```

outputs:

```
PS F:\DotNetTest> dotnet restore .\ProjectB\
  Got here before Restore!
  Restoring packages for F:\DotNetTest\ProjectB\ProjectB.csproj...
  Lock file has not changed. Skipping lock file write. Path: F:\DotNetTest\ProjectB\obj\project.assets.json
  Restore completed in 366.29 ms for F:\DotNetTest\ProjectB\ProjectB.csproj.
```

You can also use `AfterTargets` to do something after Restore.

See https://msdn.microsoft.com/en-us/library/ee216359.aspx

**Comment by atanasa** [comment]
@eerhardt As discussed in the dotnet cli issue, this did not work with the target Restore but had to use _GenerateRestoreProjectSpec. If this got fixed in the mean time, this is good. But still the issue is that Visual Studio does not get the "Got here before Restore!".

**Comment by atanasa** [comment]
@eerhardt that still does not work for `dotnet restore` on the solution.

**Comment by davkean** [comment]
Restore is not currently the correct target but we should make that work. We should mimize the target differences between command-line and VS.

**Comment by dsplaisted** [comment]
Currently, the `CollectPackageReferences` target is used to get the package references to restore in VS.  So you can do `BeforeTargets="Restore;CollectPackageReferences"` to run before restore in both VS and the command line.

It would be good to have a single target that works for building projects from the command line, building solutions from the command line, and building in VS.

**Comment by atanasa** [comment]
@dsplaisted I tried the `CollectPackageReferences` target as a before-dependency but it works more like a post-restore rather than a pre-restore target.
What I do to test it in VS is use the context menu of the solution to do "Restore NuGet Packages". The log from the Package Manager in VS Output Window shows that it is restoring the packages and that it finished. A few seconds later, my target completes. Unfortunately the log of the Package Manager is very scarce and does not seem to be affected by the MSBuild verbosity and I am not able to understand what is really happening. My custom target produces a file and this is how I know that it has completed.

To summarize, a custom target that has a `BeforeTargets=CollectPackageReferences` completes a few seconds after the log in VS reported that the restore operation has completed. If the target fails that does not fails the Restore operation either.

**Comment by atanasa** [comment]
As far as I understand, with the latest preview of dotnet cli, the restore operation is implicit, that is, you don't need to do a restore before a build operation. Is that implicit restore running as an MSBuild target? 

**Is there going to be a way to do a pre-restore target that would be triggered by this implicit restore?**

**Comment by rrelyea** [comment]
don't believe we've yet addressed this issue, even with the new noop restore work in NuGet 4.3 (shipping in VS 2017 15.3). 
Putting in 4.5 release for consideration on our backlog. Would love to get this kind of scenario solved.
@nkolev92 @emgarten

**Comment by emgarten** [comment]
I would like to see this happen also. I think to solve this restore will need to add its own target that then runs before or after CollectPackageReferences to allow this to get hit in VS. From the command line the CollectPackageReferences target is not there for all project types, so we cannot rely on only that target.

It should be a relatively simple fix, it just needs a good name and someone to try it out.

**Comment by davkean** [comment]
We should push CollectPackageReferences into NuGet.

**Comment by emgarten** [comment]
> We should push CollectPackageReferences into NuGet.

That's a good idea, is it part of the NETCore SDK currently?

**Comment by davkean** [comment]
https://github.com/dotnet/project-system/blob/708af7ef0b285f85fcc624e0b5c1a8ab193ea573/src/Targets/Microsoft.Managed.DesignTime.targets#L272

**Comment by emgarten** [comment]
Thanks @davkean! It looks like that could be copied to NuGet's targets, either the SDK or NuGet would override the other, but the result would be the same. That would make this much easier.

**Comment by atanasa** [comment]
With this approach, would it be possible to inject new PackageRefence items in that target?
The use case is - have custom tooling that calculates the PackageReference items and creates them at a time when the Restore can later see them.

**Comment by emgarten** [comment]
@atanasa that should be possible.

https://github.com/NuGet/NuGet.Client/pull/1615

**Comment by atanasa** [comment]
Any idea when this would be public? I guess this is not available in dotnet cli 2.0?

**Comment by atanasa** [comment]
@emgarten all targets that consume PackageReference items should call into this CollectPackageRefereces target, including the Publish build targets.
I've hit the case where target _ComputeExcludeFromPublishPackageReference would be build but the CollectPackageRefereces target is not called thus it does not consider all references. I've worked around this by running the "collect" target before the PrepareForPublish.

**Comment by emgarten** [comment]
For NuGet command line restore CollectPackageReferences is the only method used to get PackageReference items. Fixed with: https://github.com/NuGet/NuGet.Client/pull/1615

@atanasa I'm not familiar with `_ComputeExcludeFromPublishPackageReference` would you log an issue on the repository where that target lives?

**Comment by emgarten** [comment]
The NuGet restore changes will be part of 15.4 if it still makes it in. Otherwise 15.5.

**Comment by andreesteve** [comment]
Does this also provide a target that VS calls before restore?
Registering my target to run before CollectPackageRefereces;_GenerateRestoreProjectSpec; works well for nuget restore and msbuild /t:restore, but VS seems not to call them, or call them after it resolves other properties, like the project version I am interested in changing.

**Comment by emgarten** [comment]
VS calls CollectPackageReferences

**Comment by binki** [comment]
@emgarten @andreesteve s/CollectPackageRefereces/CollectPackageReferences/ ?

**Comment by NN---** [comment]
I added this but I don't see restoring of dotnet tools

```xml
  <Target Name="LocalToolRestore" BeforeTargets="CollectPackageRefereces">
    <Message Text="Restoring tools" Importance="High" />
    <Exec Command="dotnet tool restore" />
  </Target>

  <Target Name="LocalToolRestore2" BeforeTargets="_GenerateRestoreProjectSpec">
    <Message Text="Restoring tools" Importance="High" />
    <Exec Command="dotnet tool restore" />
  </Target>
```

**Comment by nkolev92** [comment]
@NN--- 
There's a typo there. 

It should be `CollectPackageReferences`. I updated the issue body with the correct spelling.

**Comment by Arthri** [comment]
As of writing, `CollectPackageReferences` can run during build or compilation, therefore, `CollectPackageDownloads`, `CollectFrameworkReferences`, and `CollectCentralPackageVersions` are better alternatives

**Comment by nkolev92** [comment]
> As of writing, CollectPackageReferences can run during build or compilation, therefore

I'm not sure I understand this. 

CollectPackageReferences is a restore time target, not a build or compilation.

**Comment by binki** [comment]
@nkolev92 because targets from nuget packages get run during build and assets are part of the build output, the references need to be calculated during build time even if restore is not in progress.

**Comment by nkolev92** [comment]
CollectPackageReferences is input only. It collects the declared package references and runs before restore. It is not the final list of packages that get used during build. The contract between restore & build is the assets file, not the collect target.

**Comment by Arthri** [comment]
> CollectPackageReferences is a restore time target, not a build or compilation.

CollectPackageReferences is also run during build-time, where as CollectPackageDownloads isn't

**Comment by binki** [comment]
I think I might be seeing the same thing. I did not realize it was not supposed to happen.

I have a complex solution with multiple projects. Somehow, `EntityDeploy` from `C:\Windows\Microsoft.NET\Framework\v4.0.30319\Microsoft.Data.Entity.targets`  runs on at least some dependency projects and that triggers `CollectPackageReferences`—even when I have restore disabled (I am using `msbuild` without `/r`).

My solution is not cleanly SDK projects. Not sure if that has an effect on the above.

It looks like I actually don’t care if `CollectPackageReferences` runs on build or not. I just need it to run when NuGet builds those assets.json files you were talking about. My own use case is that I have buildsystem targets which dynamically add `ProjectReference` and I had to add my own plumbing to make sure that NuGet package restore sees the transitive dependencies (for which I just have a hook against `CollectPackageReferences` which manually runs `CollectPackageReferences` against the added project dependencies because NuGet does not provide a public hook for this but instead calculates the dependency graph in its own task meaning it ignores dynamically-added `ProjectReference` and does not provide a hook making it impossible to do this correctly even though the build portion of the buildsystem supports it just fine).

**Comment by msherms2** [comment]
Old item but I see it was replied to not _too_ long ago. I don't believe this works anymore. When i rebuild via Visual Studio, I can clearly see even with the detailed log that the restore is invoked before _ANY_ targets analysis. Every combination I have tried for my situation - which is using Central Package Management, so i need to download a centralized Directory.Packages.props file - attempts restore, fails, and then invokes the build process. 

First few lines:
`Rebuild started at 1:57 PM...
Restoring NuGet packages...
To prevent NuGet from restoring packages during build, open the Visual Studio Options dialog, click on the NuGet Package Manager node and uncheck 'Allow NuGet to download missing packages during build.'
Running restore with 4 concurrent jobs.`

Only when the first project "builds" does it start analyzing Targets at all. When looking through "C:\Program Files\dotnet\sdk\8.0.204\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets", I tried a number of the targets defined there, but they end up triggering after the projects get evaluated. It's as if for a true "pre" step we need it on the solution, not in a project, but that implementation pretty much relies on an empty project to go first, which won't work for the same reason.

I'm open to ideas here but I'm fairly dead in the water. The concept of CPM is great - you update one file in one place and all your dependency tree syncs, and you just rebuild your tree. But if your tree is a lot of interdependent parts, and you need to keep your directory.packages.props file in each repository, you've just defeated that gain entirely. So we're down to just making the developers copy the file down manually or this being fixed/finding a fix, I suppose?
